Fixes #2976 - Implement feature-app-links

jhugman opened the pull request:
Fixes #2976.

This PR introduces a new feature component: `feature-app-links`.

It provides use cases for testing and opening a link in an external/non-browser app. 

Some of the UI for #2975 and the link interception in #2974 is also here, but turned off by a boolean flag.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [X] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

Amejia481 commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/RedirectDialogFragment.kt:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * A callback to trigger a download, call it when you are ready to start a download. For instance,
+     * a valid use case can be in confirmation dialog, after the positive button is clicked,
+     * this callback must be called.
```
>       * A callback to trigger a download, call it when you are ready to start a download. For instance,

See the comment above.

Amejia481 commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinkRedirect.kt:
```diff
@@ -0,0 +1,21 @@
+
+import android.content.Intent
+
+data class AppLinkRedirect(
```
Could we add some docs here?

Amejia481 commented on components/feature/app-links/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,13 @@
+  ~ file, You can obtain one at http://mozilla.org/MPL/2.0/.
+  -->
+
+<resources>
```
I think we need add this file to the `l10n.toml`file to get these strings translated.

jhugman replied:
👍

Amejia481 commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:
```diff
@@ -0,0 +1,147 @@
+
+            dialog.show(fragmentManager, FRAGMENT_TAG)
+
+            // TODO check that the user has made a choice.
```
I think we also have to take into consideration when the app gets kill by the OS, to reattach the callbacks. You can see an example on [DownloadsFeature](https://github.com/mozilla-mobile/android-components/blob/79ca1f4915b806d6292f57e527d93062ca036064/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt#L59).

jhugman replied:
:+1:

Amejia481 commented on components/feature/app-links/src/androidTest/java/mozilla/components/feature/app/links/PlaceholderTests.kt:
```diff
@@ -0,0 +1,9 @@
+
+package mozilla.components.feature.app.links
+
+class PlaceholderTests
```
Is this mean that test will come in the future ? :)

Amejia481 commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:
```diff
@@ -0,0 +1,46 @@
+ *
+ * It requires: a [Context], and a [FragmentManager].
+ */
+class AppLinksFeature(
```
I'm not quite sure if this is a feature. Normally feature starts and stops. Here we're only  exposing the interceptor and a use case @pocmo  what do you think?

Amejia481 commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:
```diff
@@ -0,0 +1,141 @@
+
+            dialog.show(fragmentManager, FRAGMENT_TAG)
+
+            // TODO check that the user has made a choice.
```
I think we should add the url issue that links to the pending `//TODO` work.

Amejia481 commented:
@jhugman  Also, I will be pretty cool 😎 , if we showcase this component in the sample browser, this will help to see how everything gets wire up together.

jhugman commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/RedirectDialogFragment.kt:
```diff
@@ -0,0 +1,46 @@
+ * This is a general representation of a dialog meant to be used in collaboration with [AppLinksFeature]
+ * to show a dialog before an external link is opened.
+ * If [SimpleRedirectDialogFragment] is not flexible enough for your use case you should inherit for this class.
+ * Be mindful to call [onConfirmRedirect] when you want to start the download.
```
Sure, this was copy pasta. Will fix.

jhugman replied:
:+1:

Amejia481 replied:
🍝

jhugman commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:
```diff
@@ -0,0 +1,147 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
```
:+1:

jhugman commented on components/feature/app-links/build.gradle:
```diff
@@ -0,0 +1,64 @@
+        exclude 'META-INF/proguard/androidx-annotations.pro'
+    }
+
+    sourceSets {
```
:+1:

jhugman commented on docs/api/package-list:
```diff
@@ -150,6 +150,7 @@ mozilla.components.concept.sync
 mozilla.components.concept.tabstray
 mozilla.components.concept.toolbar
 mozilla.components.feature.accounts
+mozilla.components.feature.app.links
```
Should this not be? What is the `package-list` for, and which components make it on to the list?

jhugman commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:
```diff
@@ -0,0 +1,147 @@
+private const val EXTRA_BROWSER_FALLBACK_URL = "browser_fallback_url"
+private const val MARKET_INTENT_URI_PACKAGE_PREFIX = "market://details?id="
+
+class AppLinksUseCases(
```
:+1:

pocmo commented on components/feature/app-links/README.md:
````diff
@@ -0,0 +1,20 @@
+```
+
+## Usage
+
````
nit: Looks like "Setting up the dependency" moved above "Usage" here :)

jhugman replied:
👍

pocmo commented on components/feature/app-links/build.gradle:
```diff
@@ -0,0 +1,61 @@
+
+    implementation Dependencies.androidx_paging
+    implementation Dependencies.androidx_lifecycle_extensions
+    kapt Dependencies.androidx_lifecycle_compiler
```
At least `paging` seems to be unused here.

jhugman replied:
:+1:

pocmo commented on components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt:
```diff
@@ -196,6 +197,13 @@ class Session(
+     * Set when a load request is received, indicating if the user was involved in the interaction.
+     */
+    var triggeredByUserInteraction: Boolean by Delegates.observable(false) {
+        _, _, new -> notifyObservers { onLoadRequest(this@Session, new) }
```
nit: Technically we are a bit "misusing" `Session` to forward events happening in `Engine.Observer`. Usually this here is a callback to notify subscribers to `Session` that a value has changed - however here we are always triggering an `onLoadRequest`.

jhugman replied:
👍

pocmo commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:
```diff
@@ -0,0 +1,124 @@
+    private val fragmentManager: FragmentManager? = null,
+    private var dialog: RedirectDialogFragment = SimpleRedirectDialogFragment.newInstance(),
+    private val useCases = AppLinksUseCases(context)
+) : SelectionAwareSessionObserver(sessionManager), LifecycleAwareFeature {
```
nit: I often prefer to implement `SelectionAwareSessionObserver` has a property instead of extending it from a feature. Extending it will expose all the methods from `SessionManager.Observer` and `Session.Observer` on this feature class - potentially confusing consumers using autocomplete.

jhugman replied:
:+1:

pocmo commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:
```diff
@@ -0,0 +1,124 @@
+ * adjoining UI. The UI will be activated in https://github.com/mozilla-mobile/android-components/issues/2974
+ * and https://github.com/mozilla-mobile/android-components/issues/2975.
+ */
+class AppLinksFeature(
```
This class doesn't seem to be covered by tests. Could it be if we mock some of the fragment stuff?

pocmo commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun showDialog(redirect: AppLinkRedirect, session: Session) {
```
The name of this method has confused me while reading the code: In some cases it doesn't actually show a dialog but instead opens the app directly. :)

jhugman replied:
:+1:

jhugman commented on components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt:
```diff
@@ -196,6 +197,13 @@ class Session(
+    /**
+     * Set when a load request is received, indicating if the user was involved in the interaction.
+     */
+    var triggeredByUserInteraction: Boolean by Delegates.observable(false) {
```
👍

jonalmeida commented:
@jhugman we've added the strings for this feature in another component for the strings freeze deadline. You can move them over to this PR later on. See https://github.com/mozilla-mobile/android-components/issues/3113

pocmo commented on components/feature/app-links/src/androidTest/java/mozilla/components/feature/app/links/PlaceholderTests.kt:
```diff
@@ -0,0 +1,9 @@
+
+package mozilla.components.feature.app.links
+
+class PlaceholderTests
```
I guess we can remove that one now? :)

jhugman replied:
:+1:

pocmo commented on components/feature/app-links/README.md:
```diff
@@ -0,0 +1,40 @@
+# [Android Components](../../../README.md) > Feature > App-Links
+
+A session components to support opening non-browser apps and `intent://` style URLs.
```
"components"

jhugman replied:
:+1:

pocmo commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinkRedirect.kt:
```diff
@@ -0,0 +1,24 @@
+    fun hasFallback() = webUrl != null && isFallback
+
+    fun isRedirect() = hasExternalApp() || hasFallback()
+}
```
I wonder if in a future iteration (not this PR) we could put this `AppLinkRedirect` into `Session` instead of a flag? Code (like the feature) could then react based on this high-level data class instead of just the flag (e.g. `AppLinksFeature` listens to a Redirect object in the Session).

jhugman commented:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1555337 to get more better reporting of a link coming out of GeckoView.

Changed `triggeredByUserInteraction` to match GeckoView docs (`triggeredByRedirect`), and added a `triggeredByWebContent`, to differentiate between the user tapping a link and entering something in the URL bar.

jhugman commented:
@pocmo @Amejia481 please could you re-review?

jhugman commented on components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:
```diff
@@ -0,0 +1,132 @@
+
+    @SuppressLint("MissingPermission")
+    @VisibleForTesting
+    fun handleRedirect(redirect: AppLinkRedirect, session: Session) {
```
:+1:
